### PR TITLE
Profile: Shuffle profile round robin thread order before taking every sample

### PR DESCRIPTION
**Pull request by NHDaly**
One approach at addressing https://github.com/JuliaLang/julia/issues/33490.

As described here: https://github.com/JuliaLang/julia/issues/9224#issuecomment-889549789, this PR has Profile.jl randomly permute the order in which it samples the threads for profiling. It still samples all threads on every pause, but it maintains an array with the order for which to sample the threads, and shuffles that array before each sample, then samples the threads according to that order.

The hope is that this will reduce the likelihood of introducing artificial contention into the program, by avoiding pausing the threads in a consistent order, which can cause a pileup of threads if they share a mutex, as described here:
https://github.com/JuliaLang/julia/issues/33490#issue-503572179

--------------------------

We seem to think that the best approach is to change the profiler to:
 - use a random sample interval
 - only sample one thread each time, chosen randomly

But this PR is a good step in the right direction.

One potential concern is that it will unacceptably reduce performance of profiling, so that remains to be seen.

I did some small measurements, and shuffling the array seems _reasonably_ fast:
```julia
julia> const x = [1,2,3]
3-element Vector{Int64}:
 1
 2
 3

julia> const seed = Ref(0)
Base.RefValue{Int64}(0)

julia> @btime @ccall jl_shuffle_int_array_inplace(pointer(x)::Ptr{Int}, length(x)::Csize_t, seed::Ptr{Int})::Cvoid
  17.172 ns (0 allocations: 0 bytes)
```
```julia
julia> const x = collect(1:100)
100-element Vector{Int64}:
   1
   ⋮
 100

julia> const seed = Ref(0)
Base.RefValue{Int64}(0)

julia> @btime @ccall jl_shuffle_int_array_inplace(pointer(x)::Ptr{Int}, length(x)::Csize_t, seed::Ptr{Int})::Cvoid
  710.507 ns (0 allocations: 0 bytes)
```

### CODE REVIEW

**Comment by NHDaly** [comment]
@vtjnash: is this more or less what you had in mind? Thanks!

**Review by vtjnash — APPROVED** [review]

**Comment by vtjnash** [comment]
Note that I don't think this will not reduce the appearance of artificial contention, since they are still likely to pile up against mutexes held while stopped for the profile, just that the contention will now be evenly distributed over threads, instead of being proportional to thread id.

**Comment by NHDaly** [comment]
I don't know if I have the power to invoke a benchmark suite run.. Can you do that? Do i just run NanoSoldier? Are there profiling-specific benchmarks to run, here?

**Comment by NHDaly** [comment]
Actually, it looks like there aren't any benchmarks for `Profile`? 🤔
https://github.com/JuliaCI/BaseBenchmarks.jl/search?q=profile

**Comment by NHDaly** [comment]
Okay excellent, it looks like there's no measurable perf impact here:

Before (Julia 1.7):
```julia
julia> Threads.nthreads()
100

julia> function myfunc()
           A = rand(200, 200, 400)
           maximum(A)
       end
myfunc (generic function with 1 method)

julia> @btime (Profile.init(n=Int(1e8), delay=0.0001); Profile.clear(); @profile myfunc())
  40.615 ms (2 allocations: 122.07 MiB)
0.9999999702097592

julia> @btime (Profile.init(n=Int(1e8), delay=0.0000001); Profile.clear(); @profile myfunc())
  42.577 ms (2 allocations: 122.07 MiB)
0.9999998284031375
```
After (this PR):
```julia
julia> @btime (Profile.init(n=Int(1e8), delay=0.0001); Profile.clear(); @profile myfunc())
  41.659 ms (2 allocations: 122.07 MiB)
0.9999999728178411

julia> @btime (Profile.init(n=Int(1e8), delay=0.0000001); Profile.clear(); @profile myfunc())
  41.263 ms (2 allocations: 122.07 MiB)
0.9999999880968238
```

I also tried with `peakflops()` with varying delay sizes, and with different numbers of threads, and similarly never saw an impact. :)

---------

So, do i just merge this now? 😬 sorry still not sure about the etiquette now that i have commit rights! 😬

**Comment by NHDaly** [comment]
Okay, after some offline confirmation, I'm merging this now!

**Comment by vtjnash** [comment]
Excellent!

**Pull request merged**
